Add query assist documentation to Observability (#6419)
opensearch-project/documentation-website: _observing-your-data/event-analytics.md

# Event analytics

> Event analytics in OpenSearch Observability provides for creating data visualizations using [Piped Processing Language]({{site.url}}{{site.baseurl}}/search-plugins/sql/ppl/index/) (PPL) queries.

Reviewer: ```suggestion
Event analytics in OpenSearch Observability allow you to create data visualizations using [Piped Processing Language]({{site.url}}{{site.baseurl}}/search-plugins/sql/ppl/index/) (PPL) queries.
```

## Getting started with event analytics

> To get started, choose **Observability** in OpenSearch Dashboards and then choose **Event analytics**. If you want to start exploring without adding your own data, choose **Add samples**. Dashboards adds sample visualizations you can interact with.

Reviewer 2: Is "Event analytics" now "Logs?"

<img width="1188" alt="Screenshot 2024-02-15 at 7 49 36 PM" src="https://github.com/opensearch-project/documentation-website/assets/105296784/46e9c188-a24b-4e7a-9ae0-3832ba2000ff">

Author: It is in this page where we talk about the log explorer for the event analytics. This is a feature to help with event analytics as it helps explore the logs.

> To get started, choose **Observability** in OpenSearch Dashboards and then choose **Logs**. If you want to start exploring without adding your own data, choose **Add samples**. Dashboards adds sample visualizations you can interact with. You can also try out pre-configured analytics in the [OpenSearch Playground](https://playground.opensearch.org/app/observability-logs#/).

Reviewer: ```suggestion
To get started, choose **Observability** in OpenSearch Dashboards and then choose **Logs**. If you want to start exploring without adding your own data, choose **Add samples**. Dashboards adds sample visualizations you can interact with. You can also try out preconfigured analytics in [OpenSearch Playground](https://playground.opensearch.org/app/observability-logs#/).
```

## Building a query

To generate custom visualizations, you must first specify a PPL query. OpenSearch Dashboards then automatically creates a visualization based on your query results.

For example, the following PPL query returns a count of how many host addresses are currently in your data.

```
source = opensearch_dashboards_sample_data_logs | fields host | stats count()
```

By default, Dashboards shows results from the last 15 minutes of your data. To see data from a different time frame, use the date and time selector to choose the desired settings.

For more information about building PPL queries, see [Piped Processing Language]({{site.url}}{{site.baseurl}}/search-plugins/sql/ppl/index).

> ### OpenSearch Query Assist

Reviewer 3: Should this be "OpenSearch Query Assistant"?

Reviewer 2: Revised throughout

This is an experimental feature and is not recommended for use in a production environment. For updates on the progress of the feature or if you want to leave feedback, see the associated [OpenSearch forum thread](https://forum.opensearch.org/t/feedback-opensearch-assistant/16741).    
{: .warning}

> To simplify query building, the **OpenSearch Assistant** toolkit offers an assistant to convert natural language queries into PPL. A screenshot is shown in the following image. 

Reviewer: ```suggestion
To simplify query building, the **OpenSearch Assistant** toolkit offers an assistant that converts natural language queries into PPL. A screenshot is shown in the following image. 
```

> ![Sample OpenSearch Query Assist View]({{site.url}}{{site.baseurl}}/images/log-explorer-query-assist.png)

Reviewer 2: ```suggestion
![Sample OpenSearch Query Assist screen view]({{site.url}}{{site.baseurl}}/images/log-explorer-query-assist.png)
```

#### Enabling Query Assistant

To enable **Query Assistant** in OpenSearch Dashboards, locate your copy of the `opensearch_dashboards.yml` file and set the following option:

```
observability.query_assist.enabled: true
observability.query_assist.ppl_agent_name: "PPL agent"
```

> To enable the option to summarize the responses, locate your copy of the `opensearch_dashboards.yml` file and set the following option:

Reviewer: "To enable summarization of responses"?

```
observability.summarize.enabled: true
observability.summarize.response_summary_agent_name: "Response summary agent"
observability.summarize.error_summary_agent_name: "Error summary agent"
```

#### Setting up Query Assistant

> To set up **Query Assistant**, follow the steps in the [Getting started guide](https://github.com/opensearch-project/dashboards-assistant/blob/main/GETTING_STARTED_GUIDE.md) on GitHub. You'll receive step-by-step setup instructions for **OpenSearch Assistant** and **Query Assistant**. To set up **Query Assistant** only, use the `query-assist-agent` template included in the guide.

Reviewer: ```suggestion
To set up **Query Assistant**, follow the steps in the [Getting started guide](https://github.com/opensearch-project/dashboards-assistant/blob/main/GETTING_STARTED_GUIDE.md) on GitHub. This guide provides step-by-step setup instructions for **OpenSearch Assistant** and **Query Assistant**. To set up **Query Assistant** only, use the `query-assist-agent` template included in the guide.
```

## Saving a visualization

> After Dashboards generates a visualization, save it if you wish to revisit it or include it in an [operational panel]({{site.url}}{{site.baseurl}}/observing-your-data/operational-panels). To save a visualization, expand the save dropdown menu to the right of the **Refresh** button, enter a name for your visualization, and then select **Save**. You can reopen saved visualizations from the event analytics page.

Reviewer: ```suggestion
After Dashboards generates a visualization, save it if you want to revisit it or include it in an [operational panel]({{site.url}}{{site.baseurl}}/observing-your-data/operational-panels). To save a visualization, expand the save dropdown menu to the right of the **Refresh** button, enter a name for your visualization, and then select **Save**. You can reopen saved visualizations on the event analytics page.
```

> After Dashboards generates a visualization, save it if you wish to revisit it or include it in an [operational panel]({{site.url}}{{site.baseurl}}/observing-your-data/operational-panels). To save a visualization, expand the save dropdown menu to the right of the **Refresh** button, enter a name for your visualization, and then select **Save**. You can reopen saved visualizations from the event analytics page.

Reviewer: Should "save" before "dropdown menu" be formatted as the name of a UI element?

## Creating event analytics visualizations and adding them to dashboards

This feature is available in OpenSearch Dashboards 2.7 and later. It works with new visualizations that use PPL to query data from OpenSearch or federated data sources such as Prometheus.
{: .note}

> Presenting your visualizations on a dashboard, instead of from the event analytics page, simplifies understanding and interpreting data at a glance. 

Reviewer: This sentence reads as a bit unnecessary. Could it be safely removed?

To create a PPL visualization, follow these steps:

1. On the main menu, choose **Visualize** > **PPL**.
2. From the **Observability** > **Logs** > **Explorer** window, enter the index source in the **PPL query** field, for example, `source = opensearch_dashboards_sample_data_flights | stats count() by DestCountry`. You must enter the query using PPL syntax. 
3. Set the time filter, for example, **This week**, and then select **Refresh**.
4. Choose the visualization type, for example, **Pie**, from the sidebar dropdown menu to the right.  

> 5. Select **Save**, and enter a name for the visualization.

Reviewer: ```suggestion
5. Select **Save** and enter a name for the visualization.
```

> You've created a new visualization that can be added to a new or an existing dashboard. To add a PPL query to a dashboard, follow these steps:  

Reviewer: ```suggestion
You've created a new visualization that can be added to a new or existing dashboard. To add a PPL query to a dashboard, follow these steps:  
```

> You've created a new visualization that can be added to a new or an existing dashboard. To add a PPL query to a dashboard, follow these steps:  

Reviewer: "You've now created"?

1. Select **Dashboards** from the main menu.
2. In the **Dashboards** window, select **Create** > **Dashboard**.
3. In the **Editing New Dashboard** window, choose **Add an existing**.
4. In the **Add panels** window, choose **PPL** from the **Types** dropdown menu, and then select the visualization. It is now displayed on your dashboard.

> 5. Select **Save**, and enter a name for the dashboard.

Reviewer: ```suggestion
5. Select **Save** and enter a name for the dashboard.
```

For more information about building PPL queries, see [Piped Processing Language]({{site.url}}{{site.baseurl}}/search-plugins/sql/ppl/index).

### OpenSearch Query Assist

This is an experimental feature and is not recommended for use in a production environment. For updates on the progress of the feature or if you want to leave feedback, see the associated [OpenSearch forum thread](https://forum.opensearch.org/t/feedback-opensearch-assistant/16741).    
{: .warning}

To make building queries easier, the OpenSearch Assistant Toolkit provides an assistant to help convert natural language queries to PPL.

![Sample OpenSearch Query Assist View]({{site.url}}{{site.baseurl}}/images/log-explorer-query-assist.png)

#### Enabling OpenSearch Query Assist
To enable **OpenSearch Query Assist** in OpenSearch Dashboards, locate your copy of the `opensearch_dashboards.yml` file and set the following option:
```
observability.query_assist.enabled: true

> observability.query_assist.ppl_agent_name: "PPL agent"

Reviewer 4: additionally, if you want to enable the summary feature on query response, set
```yml
observability.summarize.enabled: true
observability.summarize.response_summary_agent_name: "Response summary agent"
observability.summarize.error_summary_agent_name: "Error summary agent"
```

![Trace Log Correlation]({{site.url}}{{site.baseurl}}/images/trace_log_correlation.gif)

### Viewing surrounding events

If you need more information about a log event you're analyzing, you can select **View surrounding events** to gain a more comprehensive understanding of the context around the time of interest. The following demo shows this feature in action. 

![Surrounding Events]({{site.url}}{{site.baseurl}}/images/surrounding_events.gif)

### Livestreaming logs

If you prefer real-time monitoring, you can set up an interval for automatic content refresh in event analytics. Live Tail enables you to stream logs directly to OpenSearch Observability event analytics using the specified PPL query, while offering robust features like filters. This enhances your debugging process and enables seamless real-time monitoring of logs without the need for manual refreshing.

> You can select intervals and seamlessly switch between them to control the frequency of live log streaming with Live Tail. This functionality is similar to the CLI command `tail -f` command, as it retrieves only the most recent live logs, potentially eliminating a significant portion of live logs. Live Tail displays the total count of live logs received by OpenSearch during the live stream, offering insights into incoming traffic patterns. The following demo shows this feture in action. 

Reviewer 2: Is Live Tail a proper name of a feature?

![Trace Log Correlation]({{site.url}}{{site.baseurl}}/images/trace_log_correlation.gif)

### Viewing surrounding events

If you need more information about a log event you're analyzing, you can select **View surrounding events** to gain a more comprehensive understanding of the context around the time of interest. The following demo shows this feature in action. 

![Surrounding Events]({{site.url}}{{site.baseurl}}/images/surrounding_events.gif)

### Livestreaming logs

If you prefer real-time monitoring, you can set up an interval for automatic content refresh in event analytics. Live Tail enables you to stream logs directly to OpenSearch Observability event analytics using the specified PPL query, while offering robust features like filters. This enhances your debugging process and enables seamless real-time monitoring of logs without the need for manual refreshing.

## Related articles

- [Demonstrating OpenSearch Assistant toolkit for Analysis](https://www.youtube.com/watch?v=VTiJtGI2Sr4&t=152s)

> You can select intervals and seamlessly switch between them to control the frequency of live log streaming with Live Tail. This functionality is similar to the CLI command `tail -f` command, as it retrieves only the most recent live logs, potentially eliminating a significant portion of live logs. Live Tail displays the total count of live logs received by OpenSearch during the live stream, offering insights into incoming traffic patterns. The following demo shows this feature in action. 

Author: Should there be another `-` to mark the next bullet point?

Reviewer 2: Good catch. Formatting done.

### Correlating logs and traces

> If you regularly track events across applications, you can correlate logs and traces. To view the correlation, you have to index the traces according to Open Telemetry standards, similarly to [trace analytics]({{site.url}}{{site.baseurl}}/observing-your-data/trace/index/). Once you add a `TraceId` field to your logs, you can view the correlated trace information in the event explorer log details. This method correlates logs and traces that correspond to the same execution context. The following demo shows this feature in action.

Reviewer: ```suggestion
If you regularly track events across applications, you can correlate logs and traces. To view correlations, you must index the traces according to OpenTelemetry standards, similarly to [trace analytics]({{site.url}}{{site.baseurl}}/observing-your-data/trace/index/). Once you add a `TraceId` field to your logs, you can view the correlated trace information in the event explorer log details. This method correlates logs and traces that correspond to the same execution context. The following demo shows this feature in action.
```

![Trace Log Correlation]({{site.url}}{{site.baseurl}}/images/trace_log_correlation.gif)

### Viewing surrounding events

> If you need more information about a log event you're analyzing, you can select **View surrounding events** to gain a more comprehensive understanding of the context around the time of interest. The following demo shows this feature in action. 

Reviewer: Is "you're analyzing" necessary here, or could it be safely removed?

![Surrounding Events]({{site.url}}{{site.baseurl}}/images/surrounding_events.gif)

### Livestreaming logs

> If you prefer real-time monitoring, you can set up an interval for automatic content refresh in event analytics. Live Tail enables you to stream logs directly to OpenSearch Observability event analytics using the specified PPL query, while offering robust features like filters. This enhances your debugging process and enables seamless real-time monitoring of logs without the need for manual refreshing.

Reviewer: ```suggestion
If you prefer real-time monitoring, you can set up an interval for automatic content refresh in event analytics. With Live Tail, you can stream logs directly to OpenSearch Observability event analytics using the specified PPL query while leveraging robust features like filters. This can enhance your debugging process and enables seamless real-time monitoring of logs without the need for manual refreshing.
```

> If you prefer real-time monitoring, you can set up an interval for automatic content refresh in event analytics. Live Tail enables you to stream logs directly to OpenSearch Observability event analytics using the specified PPL query, while offering robust features like filters. This enhances your debugging process and enables seamless real-time monitoring of logs without the need for manual refreshing.

Reviewer: First sentence: "an interval at which event analytics content will be automatically refreshed"? Last sentence: "without the need to manually refresh content"?

> You can select intervals and seamlessly switch between them to control the frequency of live log streaming with Live Tail. This functionality is similar to the CLI command `tail -f` command, as it retrieves only the most recent live logs, potentially eliminating a significant portion of live logs. Live Tail displays the total count of live logs received by OpenSearch during the live stream, offering insights into incoming traffic patterns. The following demo shows this feature in action. 

Reviewer: ```suggestion
With Live Tail, you can select intervals and seamlessly switch between them to control the frequency of live log streaming. This functionality is similar to the `tail -f` CLI command, as it retrieves only the most recent live logs, potentially eliminating a significant portion of live logs. Live Tail displays the total number of live logs received by OpenSearch during the live stream, offering insight into incoming traffic patterns. The following demo shows this feature in action. 
```

![Live Tail]({{site.url}}{{site.baseurl}}/images/live_tail.gif)

## Related articles

> - [Demonstrating OpenSearch Assistant toolkit for Analysis](https://www.youtube.com/watch?v=VTiJtGI2Sr4&t=152s)

Reviewer: ```suggestion
- [Demonstrating the OpenSearch Assistant toolkit](https://www.youtube.com/watch?v=VTiJtGI2Sr4&t=152s)
```

> - [Demonstrating OpenSearch Assistant toolkit for Analysis](https://www.youtube.com/watch?v=VTiJtGI2Sr4&t=152s)

Reviewer: I changed to this to reflect the exact title on YouTube.
